### PR TITLE
[DTensor] decomposed sharding propagation

### PR DESCRIPTION
**Pull request by tianyu-l**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130887
* #130886

This PR adds the feature of sharding propagation via op decomposition.

#TODO: summary to be added

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

**Review by wanchaol — APPROVED** [review]
Nice work! This looks reasonably good already, only have some minor comments

File: torch/distributed/_tensor/_sharding_prop.py [hunk 1]
```diff
@@ -479,3 +513,168 @@ def _adjust_shape_and_stride_args(
...
+                possible_placements = [Replicate()] + [
+                    Shard(i) for i in range(arg_spec.ndim)
+                ]
+                possible_arg_specs = tuple(
+                    DTensorSpec(mesh, (p,), arg_spec.tensor_meta)
+                    for p in possible_placements
+                )
+                all_possible_schema.append(possible_arg_specs)
+            else:
+                all_possible_schema.append((arg_spec,))
```
[the remark this answers is not in the document]
Comment (tianyu-l) [reply torch/distributed/_tensor/_sharding_prop.py:646]: yes

File: torch/distributed/_tensor/_sharding_prop.py [hunk 2]
```diff
@@ -479,3 +513,168 @@ def _adjust_shape_and_stride_args(
...
+                    dtype=node.meta["tensor_meta"].dtype,
+                )
+                self._wrap_output_spec_tensor_meta(
+                    node.target, node_output_spec, node_output_tensor_meta
+                )
+
+                node_to_spec[node] = node_output_spec
+            elif node.op == "output":
+                output_node = node.args[0]
+                graph_output_specs = [node_to_spec[node] for node in output_node]
```
[the remark this answers is not in the document]
Comment (tianyu-l) [reply torch/distributed/_tensor/_sharding_prop.py:618]: IIRC this handles both cases of single tensor and tuple of tensors. See tests on both `aten.aminmax` (tuple of two tensors) and `aten._log_softmax` (single tensor). In other words, the `output_node` would a list of results (possibly singleton) regardless the designated output type of the function.

File: test/distributed/_tensor/test_decomposed_ops.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        x = torch.rand(8, 8, device=self.device_type)
+        softmax_dim = 1
+        y = torch.nn.functional.softmax(x, dim=softmax_dim)
+
+        # replicated input
+        x_dt = distribute_tensor(x, device_mesh, [Replicate()])
+        y_dt = torch.nn.functional.log_softmax(x_dt, dim=softmax_dim)
+
+        self.assertTrue(y_dt.placements[0].is_replicate())
+        # TODO(lty): numerical test doesn't work -- similar to the complex mul bug
```
[the remark this answers is not in the document]
Comment (dolpm) [reply test/distributed/_tensor/test_decomposed_ops.py:71]: looks like you are comparing numerics for log_softmax and regular softmax -- if they are both log this seems fine.
